### DISCUSSION TITLE
E0425: not found in this scope for name in certain rules

### DISCUSSION THREAD

**Discussion by xiaket** [pest-parser/pest#725] [Q&A] [answered]
**Describe the bug**

A same name that has been imported via super::* is available for some test cases but not for others, raising an `E0425: not found in this scope` error.

**To Reproduce**

```
# rep.pest
name = { (ASCII_ALPHA | "_") ~ (ASCII_ALPHANUMERIC | "_" | "-")* }

pipe = { command ~ ("|" ~ pipe)? }
command = { simple_command }
simple_command = { name }
```

```
# main.rs
extern crate pest;
#[macro_use]
extern crate pest_derive;

#[derive(Parser)]
#[grammar = "rep.pest"]
pub struct MyParser;


pub fn main() {
    println!("Hello");
}


#[cfg(test)]
mod tests {
    use super::*;
    use pest::parses_to;
    use pest::consumes_to;

    #[test]
    fn test_1() {
        parses_to! {
            parser: MyParser,
            input: "ls",
            rule: Rule::simple_command,
            tokens: [
                simple_command(
                    0, 2, [
                    name(0, 2)
                ])
            ]
        };
    }

    #[test]
    fn test_2() {
        parses_to! {
            parser: MyParser,
            input: "ls | wc",
            rule: Rule::pipe,
            tokens: [
                pipe( 0, 10, [
                    command( 2, 4, [
                        simple_command(2, 4, [
                            name(2, 4)
                        ]),
                    ]),
                    pipe(8, 10, [
                        command(8, 10, [
                            simple_command(8, 10 [
                                name(8, 10)
                            ])
                        ]),
                    ]),
                ])
            ]
        };
    }
}
```

**Expected behavior**

Expected behavior is both case should at least compile and run, as all the names used in `test_2` are available in the grammar file. However, when running `cargo test`, it is raising an E0425 for me:

```
error[E0425]: cannot find function `name` in this scope
  --> src/main.rs:52:33
   |
52 | ...                   name(8, 10)
   |                       ^^^^ not found in this scope
   |
help: consider importing this unit variant
   |
17 |     use crate::Rule::name;
   |
```

Which is interesting, in that if you completely remove `test_2`, `test_1` compiles and runs, and that test contains the same `name`. Moreover, if you remove the second token(starting with `pipe(8, 10)`) from `test_2`, leaving only one token(starting with `command( 2, 4`) in the tokens list, the test will compile and run(it will fail which is expected).

My suspicion is this is caused by the nested rule in the grammar file(`pipe = { command ~ ("|" ~ pipe)? }`), but I'm not that good at rust to dig it out yet.

**Additional context**

```
rustc --version
rustc 1.64.0 (a55dd71d5 2022-09-19)
```

Grateful for this wonderful tool, so much fun to play with! :)

**Comment by tomtau** [comment, marked as answer]
it seems there's a missing comma in `test2` -- when this line:
```rust
...
                            simple_command(8, 10 [
...
```

is changed to:
```rust
...
                            simple_command(8, 10, [
...
```

it will compile and run

**Comment by xiaket** [comment]
Thanks @tomtau for spotting my silly mistake! How unexpected!
